# Pandas Library
- http://pandas.pydata.org/pandas-docs/version/0.17.1/install.html
- You can find notes on the Pandas library at the link above.
- Using pandas 0.20.3 Python27


In [2]:
import csv
import numpy as np

## Import Pandas Library

In [3]:
import pandas as pd

## Read the data
- The data can be called 'Anything.csv'.  In this case it is under 'database.csv'.  We can assign a different variable name such as 'data'.

In [4]:
data = pd.read_csv('database.csv')

## Check the Data
- Next we want to check and make sure that the data is loaded.  We can just take a look at only a few rows and all the columns by using the below function.

In [5]:
data.head(20)

Institution Name  Main Office                           Branch Name  \
0   JPMorgan Chase Bank            1       JPMorgan Chase Bank Main Office   
1   JPMorgan Chase Bank            0         Vernon Hills Scarsdale Branch   
2   JPMorgan Chase Bank            0  Great Neck Northern Boulevard Branch   
3   JPMorgan Chase Bank            0                North Hartsdale Branch   
4   JPMorgan Chase Bank            0              Lawrence Rockaway Branch   
5   JPMorgan Chase Bank            0                   Mount Vernon Branch   
6   JPMorgan Chase Bank            0                    Castle Hill Branch   
7   JPMorgan Chase Bank            0                Burnside Avenue Branch   
8   JPMorgan Chase Bank            0                       Gun Hill Branch   
9   JPMorgan Chase Bank            0               Kingsbridge Road Branch   
10  JPMorgan Chase Bank            0                      Riverdale Branch   
11  JPMorgan Chase Bank            0             Washington Heights Branch   
12  JPMorgan Chase Bank            0                       Broadway Branch   
13  JPMorgan Chase Bank            0             Westchester Square Branch   
14  JPMorgan Chase Bank            0        Bay Parkway and 85th St Branch   
15  JPMorgan Chase Bank            0               Grand Army Plaza Branch   
16  JPMorgan Chase Bank            0                Carroll Gardens Branch   
17  JPMorgan Chase Bank            0     Third Avenue & 51st Street Branch   
18  JPMorgan Chase Bank            0                       East End Branch   
19  JPMorgan Chase Bank            0                        Audubon Branch   

    Branch Number Established Date Acquired Date              Street Address  \
0               0       01/01/1824           NaN        1111 Polaris Parkway   
1               2       03/20/1961           NaN       676 White Plains Road   
2               3       09/09/1963           NaN      410 Northern Boulevard   
3               4       02/19/1966           NaN    353 North Central Avenue   
4               5       01/16/1965           NaN       335 Rockaway Turnpike   
5               9       02/25/1964           NaN        22 West First Street   
6              12       12/11/1965           NaN      784 Castle Hill Avenue   
7              15       02/14/1922           NaN      5 West Burnside Avenue   
8              16       12/18/1950           NaN     1308 East Gun Hill Road   
9              18       08/18/1924           NaN    66 East Kingsbridge Road   
10             20       06/01/1950           NaN       3775 Riverdale Avenue   
11             21       01/02/1909           NaN  1421 Saint Nicholas Avenue   
12             22       07/12/1926           NaN               5581 Broadway   
13             24       06/29/1959           NaN    2725 East Tremont Avenue   
14             26       03/25/1964           NaN            8418 Bay Parkway   
15             33       03/14/1927           NaN         401 Flatbush Avenue   
16             34       06/09/1913           NaN          79 Hamilton Avenue   
17             37       05/05/1961           NaN            850 Third Avenue   
18             38       05/22/1961           NaN            1495 York Avenue   
19             40       04/02/1917           NaN               3940 Broadway   

             City       County State  Zipcode  Latitude  Longitude  \
0        Columbus     Delaware    OH    43240  40.14453  -82.99115   
1       Scarsdale  Westchester    NY    10583  40.97008  -73.80670   
2      Great Neck       Nassau    NY    11021  40.77944  -73.72240   
3       Hartsdale  Westchester    NY    10530  41.02654  -73.79168   
4        Lawrence       Nassau    NY    11559  40.62715  -73.73675   
5    Mount Vernon  Westchester    NY    10550  40.91144  -73.83804   
6           Bronx        Bronx    NY    10473  40.82292  -73.84887   
7           Bronx        Bronx    NY    10453  40.85389  -73.90760   
8           Bronx        Bronx    NY    10469  40.87151  

- We can also check the end/tail of the data which in this case is the last five rows as shown below:

In [6]:
data.tail()

Institution Name  Main Office                      Branch Name  \
5408  JPMorgan Chase Bank            0  Alameda and Compton Blvd Branch   
5409  JPMorgan Chase Bank            0            Charleston and Rancho   
5410  JPMorgan Chase Bank            0                Los Olivos Branch   
5411  JPMorgan Chase Bank            0               Lake Forest Branch   
5412  JPMorgan Chase Bank            0                Buffalo-Mm Branch   

      Branch Number Established Date Acquired Date            Street Address  \
5408           7981       11/03/2015           NaN        251 E Compton Blvd   
5409           7982       02/02/2016           NaN   2311 W Chafrleston Blvd   
5410           7984       03/15/2016           NaN  8593 Irvine Center Drive   
5411           7988       01/01/2016           NaN            5660 Read Blvd   
5412           7989       01/01/2016           NaN           350 Main Street   

             City       County State  Zipcode  Latitude  Longitude  \
5408      Compton  Los Angeles    CA    90220  33.89601 -118.22084   
5409    Las Vegas        Clark    NV    89102  36.14270 -115.19043   
5410       Irvine       Orange    CA    92618  33.64412 -117.74453   
5411  New Orleans      Orleans    LA    70127  30.03205  -89.97260   
5412      Buffalo         Erie    NY    14202  42.88429  -78.87487   

      2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
5408            NaN            NaN            NaN            NaN   
5409            NaN            NaN            NaN            NaN   
5410            NaN            NaN            NaN            NaN   
5411            NaN            NaN            NaN            NaN   
5412            NaN            NaN            NaN            NaN   

      2014 Deposits  2015 Deposits  2016 Deposits  
5408            NaN            NaN           3446  
5409            NaN            NaN           2666  
5410            NaN            NaN           6689  
5411            NaN            NaN          94133  
5412            NaN            NaN          45596

## Dimensions
- In order to create an idea it is better to check how many rows and columns the data has.  The first number is for the rows and the second if for the columns.

In [7]:
data.shape

(5413, 20)

## Check the Type of the Data

In [8]:
type(data)

pandas.core.frame.DataFrame

## Set Index
- Column named 'State' is set as the index column.

In [9]:
df1 = data.set_index('State')

df1

Institution Name  Main Office  \
State                                     
OH     JPMorgan Chase Bank            1   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
...                    ...          ...   
FL     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
IL     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CT     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
NJ     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CO     JPMorgan Chase Bank            0   
IL     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
OK     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
FL     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
NV     JPMorgan Chase Bank            0   
CA     JPMorgan Chase Bank            0   
LA     JPMorgan Chase Bank            0   
NY     JPMorgan Chase Bank            0   

                                       Branch Name  Branch Number  \
State                                                               
OH                 JPMorgan Chase Bank Main Office              0   
NY                   Vernon Hills Scarsdale Branch              2   
NY            Great Neck Northern Boulevard Branch              3   
NY                          North Hartsdale Branch              4   
NY                        Lawrence Rockaway Branch              5   
NY                             Mount Vernon Branch              9   
NY                              Castle Hill Branch             12   
NY                          Burnside Avenue Branch             15   
NY                                 Gun Hill Branch             16   
NY                         Kingsbridge Road Branch             18   
NY                                Riverdale Branch             20   
NY                       Washington Heights Branch             21   
NY                                 Broadway Branch             22   
NY                       Westchester Square Branch             24   
NY                  Bay Parkway and 85th St Branch             26   
NY                         Grand Army Plaza Branch             33   
NY                          Carroll Gardens Branch        

## Group All Rows Accordingly to State
- We can aggregate and ask for the sum total of State.  Or aggregate and ask for the mean.
- The function below aggregates all the data that corresponds to each State into one single row per State.  For example, if AZ has 256 rows out of 5,413 total rows, now it will all show as one mean of all the data into one single row.  Dataframe has 5,413 rows and we get 26 total rows after running the function below.

In [10]:
df2 = df1.groupby('State').mean()

df2

Main Office  Branch Number       Zipcode   Latitude   Longitude  \
State                                                                    
AZ        0.000000    3633.179688  85398.687500  33.383542 -111.899391   
CA        0.000000    5913.613161  92805.446660  35.336353 -119.409642   
CO        0.000000    4774.482456  80312.236842  39.755065 -104.976522   
CT        0.000000    3319.280000   6753.000000  41.187905  -73.375007   
DC        0.000000    4409.000000  20006.000000  38.900240  -77.038070   
FL        0.000000    6309.313131  33396.648990  27.123376  -81.014857   
GA        0.000000    6613.759036  30164.012048  33.892526  -84.334768   
ID        0.000000    6381.050000  83621.650000  43.854341 -115.635778   
IL        0.000000    3957.638814  60467.326146  41.400228  -87.017360   
IN        0.000000    3155.175141  46546.960452  40.247763  -86.238464   
KY        0.000000    3091.290323  40628.467742  38.144717  -85.291730   
LA        0.000000    2338.798701  70635.077922  30.629858  -91.516504   
MA        0.000000    4407.000000   2110.000000  42.355970  -71.049810   
MI        0.000000    3024.394309  48530.886179  42.791790  -83.849739   
NJ        0.000000    4513.926606   7700.325688  40.697085  -74.280608   
NV        0.000000    6434.812500  89100.708333  36.122835 -115.177113   
NY        0.000000    3011.433246  11139.099476  40.968160  -73.964581   
OH        0.003509    2106.245614  44291.147368  40.376919  -82.660154   
OK        0.000000    3428.093750  73409.968750  35.671267  -97.014822   
OR        0.000000    5966.743363  97275.681416  44.726245 -122.713529   
PA        0.000000    4406.000000  19103.000000  39.953090  -75.174020   
TX        0.000000    3823.534848  76838.187879  31.118924  -96.833499   
UT        0.000000    4602.203704  84194.203704  40.442665 -111.975328   
WA        0.000000    5894.200980  98366.514706  47.394824 -121.753010   
WI        0.000000    2896.148649  53671.797297  43.378328  -88.352818   
WV        0.000000    1809.730769  25662.461538  38.511245  -81.466225   

       2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
AZ      7.589532e+04   7.717828e+04   8.357267e+04   9.196123e+04   
CA      8.322127e+04   8.205084e+04   8.260358e+04   8.545739e+04   
CO      6.928259e+04   8.152999e+04   8.024558e+04   9.589530e+04   
CT      8.899461e+04   9.124488e+04   9.220400e+04   9.616670e+04   
DC      0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
FL      4.433210e+04   4.721947e+04   4.885668e+04   5.250518e+04   
GA      1.170818e+04   1.301828e+04   1.691926e+04   2.125203e+04   
ID      1.914411e+04   1.909026e+04   2.031915e+04   2.137000e+04   
IL      1.457416e+05   1.685098e+05   1.951250e+05   2.091494e+05   
IN      7.072699e+04   7.101410e+04   7.576674e+04   8.109440e+04   
KY      6.736108e+04   7.177548e+04   8.162852e+04   7.552977e+04   
LA      8.927389e+04   9.189092e+04   9.970439e+04   1.061476e+05   
MA      0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
MI      8.434725e+04   1.011638e+05   1.150547e+05   1.263822e+05   
NJ      4.938942e+04   5.584091e+04   5.738146e+04   6.551845e+04   
NV      2.934353e+04   3.179887e+04   3.645212e+04   4.002855e+04   
NY      3.849342e+05   4.713859e+05   5.245762e+05   5.651023e+05   
OH      2.347836e+06   2.744006e+06   3.015689e+06   3.306157e+06   
OK      5.154726e+04   5.202906e+04   5.638845e+04   6.025790e+04   
OR      3.968529e+04   3.893520e+04   3.856117e+04   3.927763e+04   
PA      0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
TX      1.277518e+05   1.513263e+05   1.626560e+05   1.905592e+05   
UT      1.267652e+05   1.240244e+05   6.785041e+04   8.165352e+04   
WA      4.865160e+04   4.808073e+04   5.044664e+04   5.443459e+04   
WI      7.746509e+04   9.133796e+04   9.855585e+04   1.067444e+05   
WV      6.198619e+04   6.044546e+04   6.479338e+04   6.657346e+04   

 

## Second Way of Aggregating
- Below is the example of grouping and asking for the sum instead of the mean().

In [11]:
df2 = df1.groupby('State').sum()

df2

Main Office  Branch Number   Zipcode     Latitude     Longitude  \
State                                                                    
AZ               0         930094  21862064   8512.80314  -28534.34458   
CA               0        5931354  93083863  34629.62581 -117021.44869   
CO               0         544291   9155595   4532.07743  -11967.32348   
CT               0         165964    337650   1977.01942   -3522.00035   
DC               0           4409     20006     38.90024     -77.03807   
FL               0        2498488  13225073  10279.75942  -30704.63079   
GA               0         548942   2503613   2711.40211   -6746.78145   
ID               0         127621   1672433    877.08682   -2312.71556   
IL               0        1468284  22433378  15276.68427  -32109.40570   
IN               0         558466   8238812   7123.85408  -15264.20818   
KY               0         191660   2518965   2364.97248   -5288.08726   
LA               0         360175  10877802   4686.36827  -14002.02517   
MA               0           4407      2110     42.35597     -71.04981   
MI               0         744001  11938598  10526.78037  -20627.03587   
NJ               0         984036   1678671   8871.96447  -16193.17263   
NV               0         308871   4276834   1661.65039   -5298.14719   
NY               0        2300735   8510272  30889.99275  -55769.29402   
OH               1         600280  12622977  11467.04513  -23475.48377   
OK               0         109699   2349119   1141.48053   -3104.47429   
OR               0         674242  10992152   5009.33947  -13743.91520   
PA               0           4406     19103     39.95309     -75.17402   
TX               0        2523533  50713204  20445.13331  -63619.60887   
UT               0         248519   4546487   2183.90392   -6046.66770   
WA               0        1202417  20066769   9668.54413  -24837.61408   
WI               0         214315   3971713   3209.99624   -6538.10854   
WV               0          47053    667224   1001.29237   -2118.12184   

       2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
AZ        18518457.0     19217392.0     21227459.0     23358152.0   
CA        59586426.0     63507350.0     71452096.0     80244493.0   
CO         6997542.0      8642179.0      8827014.0     10836169.0   
CT         3648779.0      3832285.0      4241384.0      4519835.0   
DC               0.0            0.0            0.0            0.0   
FL         9708730.0     11568770.0     14901287.0     18114286.0   
GA          573701.0       833170.0      1235106.0      1657658.0   
ID          363738.0       362715.0       406383.0       427400.0   
IL        47511768.0     56113765.0     68488872.0     74457178.0   
IN        12447950.0     12569495.0     13410713.0     14353708.0   
KY         4176387.0      4450080.0      5060968.0      4682846.0   
LA        13569631.0     13967420.0     15155067.0     16134439.0   
MA               0.0            0.0            0.0            0.0   
MI        20580730.0     24785129.0     28303465.0     31090021.0   
NJ         9532159.0     11112342.0     11877962.0     13824392.0   
NV          997680.0      1240156.0      1567441.0      1881342.0   
NY       281001947.0    348825565.0    390809240.0    422696548.0   
OH       650350577.0    762833720.0    853440012.0    938948496.0   
OK         1597965.0      1612901.0      1748042.0      1867995.0   
OR         4008214.0      4010326.0      4241729.0      4438372.0   
PA               0.0            0.0            0.0            0.0   
TX        79461631.0     95184270.0    105401093.0    125006852.0   
UT         6084730.0      6325245.0      3663922.0      4409290.0   
WA         9000546.0      9327661.0     10038882.0     10995788.0   
WI         5732417.0      6759009.0      7293133.0      7899084.0   
WV         1611641.0      1571582.0      1684628.0      1730910.0   

 

## Select Desired Columns
- The columns 'Branch Number', 'Zipcode', 'Latitude',and  'Longitude' are not useful anymore.  All the zipcodes are sumed up and meaningless to use.  Same for the mentioned columns as well.  But we can drop and get rid of those columns and keep what we need using the below example.

In [12]:
df3 = df2.iloc[:, 5:]

df3

2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
AZ        18518457.0     19217392.0     21227459.0     23358152.0   
CA        59586426.0     63507350.0     71452096.0     80244493.0   
CO         6997542.0      8642179.0      8827014.0     10836169.0   
CT         3648779.0      3832285.0      4241384.0      4519835.0   
DC               0.0            0.0            0.0            0.0   
FL         9708730.0     11568770.0     14901287.0     18114286.0   
GA          573701.0       833170.0      1235106.0      1657658.0   
ID          363738.0       362715.0       406383.0       427400.0   
IL        47511768.0     56113765.0     68488872.0     74457178.0   
IN        12447950.0     12569495.0     13410713.0     14353708.0   
KY         4176387.0      4450080.0      5060968.0      4682846.0   
LA        13569631.0     13967420.0     15155067.0     16134439.0   
MA               0.0            0.0            0.0            0.0   
MI        20580730.0     24785129.0     28303465.0     31090021.0   
NJ         9532159.0     11112342.0     11877962.0     13824392.0   
NV          997680.0      1240156.0      1567441.0      1881342.0   
NY       281001947.0    348825565.0    390809240.0    422696548.0   
OH       650350577.0    762833720.0    853440012.0    938948496.0   
OK         1597965.0      1612901.0      1748042.0      1867995.0   
OR         4008214.0      4010326.0      4241729.0      4438372.0   
PA               0.0            0.0            0.0            0.0   
TX        79461631.0     95184270.0    105401093.0    125006852.0   
UT         6084730.0      6325245.0      3663922.0      4409290.0   
WA         9000546.0      9327661.0     10038882.0     10995788.0   
WI         5732417.0      6759009.0      7293133.0      7899084.0   
WV         1611641.0      1571582.0      1684628.0      1730910.0   

       2014 Deposits  2015 Deposits  2016 Deposits  
State                                               
AZ      2.476008e+07   2.735852e+07       30126930  
CA      9.134881e+07   1.022120e+08      118958728  
CO      9.749824e+06   1.018713e+07       11528794  
CT      4.802183e+06   5.003594e+06        5533022  
DC      0.000000e+00   0.000000e+00              0  
FL      2.135011e+07   2.474166e+07       28836780  
GA      2.187653e+06   2.644146e+06        3179306  
ID      4.987720e+05   5.778230e+05         673964  
IL      8.707051e+07   8.672616e+07       84405502  
IN      1.571671e+07   1.682736e+07       17887763  
KY      5.092516e+06   5.521501e+06        5667540  
LA      1.674408e+07   1.754247e+07       17589216  
MA      0.000000e+00   0.000000e+00              0  
MI      3.605758e+07   3.898354e+07       42006213  
NJ      1.545324e+07   1.706048e+07       19249470  
NV      2.250576e+06   2.530200e+06        2909706  
NY      4.596692e+08   4.787386e+08      536752799  
OH      1.056778e+09   1.094947e+09     1183589468  
OK      2.128120e+06   2.126367e+06        4857743  
OR      4.863094e+06   5.350523e+06        6099453  
PA      0.000000e+00   0.000000e+00              0  
TX      1.631469e+08   1.469289e+08      151393753  
UT      4.257220e+06   1.210851e+07       12709215  
WA      1.216092e+07   1.340261e+07       15191659  
WI      8.466207e+06   9.358616e+06        9410758  
WV      1.742605e+06   1.752744e+06        1812236

## Note from the Above Example
- The first : inside square brackets denotes all the rows, and the second : after number 5 denotes all the columns after the fifth column.  So, instead we can try [:, 5:6] which means all the rows and only the fifth and sixth column are to be selected from the entire dataframe.
- Same goes for the rows.  So, at the first : we can place a number in front or after like this [4:, 5:] which will return all the rows after the fourth row (excluding the first four).  Or the opposite [:4, 5:] which will return only the first 4 rows.

## Method describe()
- Computes common aggregates for each column and returns the result.  It is an easy and short method that provides a lot of information.  It summarizes the data and gives understanding of what values we have, what story do we see, and what can we do and show.

In [13]:
df3.describe()

2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
count   2.600000e+01   2.600000e+01   2.600000e+01   2.600000e+01   
mean    4.796397e+07   5.648664e+07   6.324907e+07   6.975289e+07   
std     1.350215e+08   1.598470e+08   1.789220e+08   1.963973e+08   
min     0.000000e+00   0.000000e+00   0.000000e+00   0.000000e+00   
25%     1.601384e+06   1.581912e+06   1.700482e+06   1.871332e+06   
50%     6.541136e+06   7.700594e+06   8.060074e+06   9.367626e+06   
75%     1.728125e+07   1.790490e+07   1.970936e+07   2.204719e+07   
max     6.503506e+08   7.628337e+08   8.534400e+08   9.389485e+08   

       2014 Deposits  2015 Deposits  2016 Deposits  
count   2.600000e+01   2.600000e+01   2.600000e+01  
mean    7.870365e+07   8.163962e+07   8.886039e+07  
std     2.202840e+08   2.279255e+08   2.476405e+08  
min     0.000000e+00   0.000000e+00   0.000000e+00  
25%     2.203384e+06   2.558686e+06   3.598915e+06  
50%     9.108016e+06   1.114782e+07   1.211900e+07  
75%     2.390759e+07   2.670431e+07   2.980439e+07  
max     1.056778e+09   1.094947e+09   1.183589e+09

## Get Rid of Rows
- If you look at the above data, rows that correspond to 'DC', 'PA', & 'MA' have 0 values.  We can get rid of those rows and have a better data frame to use by doing the following:

In [14]:
df4 = df3.drop(['DC', 'MA', 'PA'])
df4

2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
AZ        18518457.0     19217392.0     21227459.0     23358152.0   
CA        59586426.0     63507350.0     71452096.0     80244493.0   
CO         6997542.0      8642179.0      8827014.0     10836169.0   
CT         3648779.0      3832285.0      4241384.0      4519835.0   
FL         9708730.0     11568770.0     14901287.0     18114286.0   
GA          573701.0       833170.0      1235106.0      1657658.0   
ID          363738.0       362715.0       406383.0       427400.0   
IL        47511768.0     56113765.0     68488872.0     74457178.0   
IN        12447950.0     12569495.0     13410713.0     14353708.0   
KY         4176387.0      4450080.0      5060968.0      4682846.0   
LA        13569631.0     13967420.0     15155067.0     16134439.0   
MI        20580730.0     24785129.0     28303465.0     31090021.0   
NJ         9532159.0     11112342.0     11877962.0     13824392.0   
NV          997680.0      1240156.0      1567441.0      1881342.0   
NY       281001947.0    348825565.0    390809240.0    422696548.0   
OH       650350577.0    762833720.0    853440012.0    938948496.0   
OK         1597965.0      1612901.0      1748042.0      1867995.0   
OR         4008214.0      4010326.0      4241729.0      4438372.0   
TX        79461631.0     95184270.0    105401093.0    125006852.0   
UT         6084730.0      6325245.0      3663922.0      4409290.0   
WA         9000546.0      9327661.0     10038882.0     10995788.0   
WI         5732417.0      6759009.0      7293133.0      7899084.0   
WV         1611641.0      1571582.0      1684628.0      1730910.0   

       2014 Deposits  2015 Deposits  2016 Deposits  
State                                               
AZ      2.476008e+07   2.735852e+07       30126930  
CA      9.134881e+07   1.022120e+08      118958728  
CO      9.749824e+06   1.018713e+07       11528794  
CT      4.802183e+06   5.003594e+06        5533022  
FL      2.135011e+07   2.474166e+07       28836780  
GA      2.187653e+06   2.644146e+06        3179306  
ID      4.987720e+05   5.778230e+05         673964  
IL      8.707051e+07   8.672616e+07       84405502  
IN      1.571671e+07   1.682736e+07       17887763  
KY      5.092516e+06   5.521501e+06        5667540  
LA      1.674408e+07   1.754247e+07       17589216  
MI      3.605758e+07   3.898354e+07       42006213  
NJ      1.545324e+07   1.706048e+07       19249470  
NV      2.250576e+06   2.530200e+06        2909706  
NY      4.596692e+08   4.787386e+08      536752799  
OH      1.056778e+09   1.094947e+09     1183589468  
OK      2.128120e+06   2.126367e+06        4857743  
OR      4.863094e+06   5.350523e+06        6099453  
TX      1.631469e+08   1.469289e+08      151393753  
UT      4.257220e+06   1.210851e+07       12709215  
WA      1.216092e+07   1.340261e+07       15191659  
WI      8.466207e+06   9.358616e+06        9410758  
WV      1.742605e+06   1.752744e+06        1812236

## Select Only One Row
- Select only one state, in this case WV, and show all its variables/ columns.

In [15]:
df4.loc['WV']

2010 Deposits    1611641.0
2011 Deposits    1571582.0
2012 Deposits    1684628.0
2013 Deposits    1730910.0
2014 Deposits    1742605.0
2015 Deposits    1752744.0
2016 Deposits    1812236.0
Name: WV, dtype: float64

## Select Only a Certain Row
- Example below selects only the rows that belong to Arizona.  There are 256 rows and 20 columns out of 5,413 that match with Arizona.

In [16]:
df5 = data[data.State == 'AZ']
df5

Institution Name  Main Office  \
1234  JPMorgan Chase Bank            0   
1235  JPMorgan Chase Bank            0   
1236  JPMorgan Chase Bank            0   
1237  JPMorgan Chase Bank            0   
1238  JPMorgan Chase Bank            0   
1239  JPMorgan Chase Bank            0   
1240  JPMorgan Chase Bank            0   
1241  JPMorgan Chase Bank            0   
1242  JPMorgan Chase Bank            0   
1243  JPMorgan Chase Bank            0   
1244  JPMorgan Chase Bank            0   
1245  JPMorgan Chase Bank            0   
1246  JPMorgan Chase Bank            0   
1247  JPMorgan Chase Bank            0   
1248  JPMorgan Chase Bank            0   
1249  JPMorgan Chase Bank            0   
1250  JPMorgan Chase Bank            0   
1251  JPMorgan Chase Bank            0   
1252  JPMorgan Chase Bank            0   
1253  JPMorgan Chase Bank            0   
1254  JPMorgan Chase Bank            0   
1255  JPMorgan Chase Bank            0   
1256  JPMorgan Chase Bank            0   
1257  JPMorgan Chase Bank            0   
1258  JPMorgan Chase Bank            0   
1259  JPMorgan Chase Bank            0   
1260  JPMorgan Chase Bank            0   
1261  JPMorgan Chase Bank            0   
1262  JPMorgan Chase Bank            0   
1263  JPMorgan Chase Bank            0   
...                   ...          ...   
4423  JPMorgan Chase Bank            0   
4424  JPMorgan Chase Bank            0   
4425  JPMorgan Chase Bank            0   
4426  JPMorgan Chase Bank            0   
4427  JPMorgan Chase Bank            0   
4429  JPMorgan Chase Bank            0   
4537  JPMorgan Chase Bank            0   
4538  JPMorgan Chase Bank            0   
4539  JPMorgan Chase Bank            0   
4569  JPMorgan Chase Bank            0   
4570  JPMorgan Chase Bank            0   
4573  JPMorgan Chase Bank            0   
4617  JPMorgan Chase Bank            0   
4621  JPMorgan Chase Bank            0   
4630  JPMorgan Chase Bank            0   
4631  JPMorgan Chase Bank            0   
4647  JPMorgan Chase Bank            0   
4652  JPMorgan Chase Bank            0   
4679  JPMorgan Chase Bank            0   
4686  JPMorgan Chase Bank            0   
4700  JPMorgan Chase Bank            0   
4701  JPMorgan Chase Bank            0   
4760  JPMorgan Chase Bank            0   
4871  JPMorgan Chase Bank            0   
4921  JPMorgan Chase Bank            0   
4973  JPMorgan Chase Bank            0   
4974  JPMorgan Chase Bank            0   
5033  JPMorgan Chase Bank            0   
5249  JPMorgan Chase Bank            0   
5267  JPMorgan Chase Bank            0   

                                        Branch Name  Branch Number  \
1234                             Chase Tower Branch           2414   
1235                                 Buckeye Branch           2415   
1236                                Carefree Branch           2416   
1237                           Glendale Main Branch           2419   
1238                            Grand Canyon Branch           2420   
1239                             Lake Havasu Branch           2421   
1240               Camelback and 59th Avenue Branch           2422   
1241                 Mcdowelland 74th Street Branch           2423   
1242                               Mesa Main Branch           2424   
1243                                  Payson Branch           2426   
1244                Camelback And16th Street Branch           2427   
1245           Indian School and 33rd Avenue Branch           2428   
1246                  Buckeye And16th Street Branch           2429   
1247                    Central and Columbus Branch           2430   
1248                 Glendale and 7th Street Branch           2431   
1249            Bethany Home and 19th Avenue Branch           2432   
1250                          West Van Buren Branch           2433   
1251               Camelback and 44th Street Branch           2434   
1252                  Thomas and 24th Street Branch           2435   
1253

## Rank Data (Ascending Order)
- Rank the data in ascending order.  The example below lists the smallest number first by specifying (..ascending=True).  We get the same result using (...ascending=1).

In [17]:
ranked1 = df4.sort_values('2010 Deposits', ascending=True)
ranked1

2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
ID          363738.0       362715.0       406383.0       427400.0   
GA          573701.0       833170.0      1235106.0      1657658.0   
NV          997680.0      1240156.0      1567441.0      1881342.0   
OK         1597965.0      1612901.0      1748042.0      1867995.0   
WV         1611641.0      1571582.0      1684628.0      1730910.0   
CT         3648779.0      3832285.0      4241384.0      4519835.0   
OR         4008214.0      4010326.0      4241729.0      4438372.0   
KY         4176387.0      4450080.0      5060968.0      4682846.0   
WI         5732417.0      6759009.0      7293133.0      7899084.0   
UT         6084730.0      6325245.0      3663922.0      4409290.0   
CO         6997542.0      8642179.0      8827014.0     10836169.0   
WA         9000546.0      9327661.0     10038882.0     10995788.0   
NJ         9532159.0     11112342.0     11877962.0     13824392.0   
FL         9708730.0     11568770.0     14901287.0     18114286.0   
IN        12447950.0     12569495.0     13410713.0     14353708.0   
LA        13569631.0     13967420.0     15155067.0     16134439.0   
AZ        18518457.0     19217392.0     21227459.0     23358152.0   
MI        20580730.0     24785129.0     28303465.0     31090021.0   
IL        47511768.0     56113765.0     68488872.0     74457178.0   
CA        59586426.0     63507350.0     71452096.0     80244493.0   
TX        79461631.0     95184270.0    105401093.0    125006852.0   
NY       281001947.0    348825565.0    390809240.0    422696548.0   
OH       650350577.0    762833720.0    853440012.0    938948496.0   

       2014 Deposits  2015 Deposits  2016 Deposits  
State                                               
ID      4.987720e+05   5.778230e+05         673964  
GA      2.187653e+06   2.644146e+06        3179306  
NV      2.250576e+06   2.530200e+06        2909706  
OK      2.128120e+06   2.126367e+06        4857743  
WV      1.742605e+06   1.752744e+06        1812236  
CT      4.802183e+06   5.003594e+06        5533022  
OR      4.863094e+06   5.350523e+06        6099453  
KY      5.092516e+06   5.521501e+06        5667540  
WI      8.466207e+06   9.358616e+06        9410758  
UT      4.257220e+06   1.210851e+07       12709215  
CO      9.749824e+06   1.018713e+07       11528794  
WA      1.216092e+07   1.340261e+07       15191659  
NJ      1.545324e+07   1.706048e+07       19249470  
FL      2.135011e+07   2.474166e+07       28836780  
IN      1.571671e+07   1.682736e+07       17887763  
LA      1.674408e+07   1.754247e+07       17589216  
AZ      2.476008e+07   2.735852e+07       30126930  
MI      3.605758e+07   3.898354e+07       42006213  
IL      8.707051e+07   8.672616e+07       84405502  
CA      9.134881e+07   1.022120e+08      118958728  
TX      1.631469e+08   1.469289e+08      151393753  
NY      4.596692e+08   4.787386e+08      536752799  
OH      1.056778e+09   1.094947e+09     1183589468

## Rank Data (Descending Order)
- Rank the data in descending order.  The example below lists the biggest number first by specifying (...ascending=False) same as (...ascending=0).

In [18]:
ranked2 = df4.sort_values('2011 Deposits', ascending=False)
ranked2

2010 Deposits  2011 Deposits  2012 Deposits  2013 Deposits  \
State                                                               
OH       650350577.0    762833720.0    853440012.0    938948496.0   
NY       281001947.0    348825565.0    390809240.0    422696548.0   
TX        79461631.0     95184270.0    105401093.0    125006852.0   
CA        59586426.0     63507350.0     71452096.0     80244493.0   
IL        47511768.0     56113765.0     68488872.0     74457178.0   
MI        20580730.0     24785129.0     28303465.0     31090021.0   
AZ        18518457.0     19217392.0     21227459.0     23358152.0   
LA        13569631.0     13967420.0     15155067.0     16134439.0   
IN        12447950.0     12569495.0     13410713.0     14353708.0   
FL         9708730.0     11568770.0     14901287.0     18114286.0   
NJ         9532159.0     11112342.0     11877962.0     13824392.0   
WA         9000546.0      9327661.0     10038882.0     10995788.0   
CO         6997542.0      8642179.0      8827014.0     10836169.0   
WI         5732417.0      6759009.0      7293133.0      7899084.0   
UT         6084730.0      6325245.0      3663922.0      4409290.0   
KY         4176387.0      4450080.0      5060968.0      4682846.0   
OR         4008214.0      4010326.0      4241729.0      4438372.0   
CT         3648779.0      3832285.0      4241384.0      4519835.0   
OK         1597965.0      1612901.0      1748042.0      1867995.0   
WV         1611641.0      1571582.0      1684628.0      1730910.0   
NV          997680.0      1240156.0      1567441.0      1881342.0   
GA          573701.0       833170.0      1235106.0      1657658.0   
ID          363738.0       362715.0       406383.0       427400.0   

       2014 Deposits  2015 Deposits  2016 Deposits  
State                                               
OH      1.056778e+09   1.094947e+09     1183589468  
NY      4.596692e+08   4.787386e+08      536752799  
TX      1.631469e+08   1.469289e+08      151393753  
CA      9.134881e+07   1.022120e+08      118958728  
IL      8.707051e+07   8.672616e+07       84405502  
MI      3.605758e+07   3.898354e+07       42006213  
AZ      2.476008e+07   2.735852e+07       30126930  
LA      1.674408e+07   1.754247e+07       17589216  
IN      1.571671e+07   1.682736e+07       17887763  
FL      2.135011e+07   2.474166e+07       28836780  
NJ      1.545324e+07   1.706048e+07       19249470  
WA      1.216092e+07   1.340261e+07       15191659  
CO      9.749824e+06   1.018713e+07       11528794  
WI      8.466207e+06   9.358616e+06        9410758  
UT      4.257220e+06   1.210851e+07       12709215  
KY      5.092516e+06   5.521501e+06        5667540  
OR      4.863094e+06   5.350523e+06        6099453  
CT      4.802183e+06   5.003594e+06        5533022  
OK      2.128120e+06   2.126367e+06        4857743  
WV      1.742605e+06   1.752744e+06        1812236  
NV      2.250576e+06   2.530200e+06        2909706  
GA      2.187653e+06   2.644146e+06        3179306  
ID      4.987720e+05   5.778230e+05         673964

## Add Column
- We can add a column in the dataframe and also rank it.

In [19]:
df4['rank_column'] = df4['2010 Deposits'].rank(ascending=False)
data

Institution Name  Main Office  \
0     JPMorgan Chase Bank            1   
1     JPMorgan Chase Bank            0   
2     JPMorgan Chase Bank            0   
3     JPMorgan Chase Bank            0   
4     JPMorgan Chase Bank            0   
5     JPMorgan Chase Bank            0   
6     JPMorgan Chase Bank            0   
7     JPMorgan Chase Bank            0   
8     JPMorgan Chase Bank            0   
9     JPMorgan Chase Bank            0   
10    JPMorgan Chase Bank            0   
11    JPMorgan Chase Bank            0   
12    JPMorgan Chase Bank            0   
13    JPMorgan Chase Bank            0   
14    JPMorgan Chase Bank            0   
15    JPMorgan Chase Bank            0   
16    JPMorgan Chase Bank            0   
17    JPMorgan Chase Bank            0   
18    JPMorgan Chase Bank            0   
19    JPMorgan Chase Bank            0   
20    JPMorgan Chase Bank            0   
21    JPMorgan Chase Bank            0   
22    JPMorgan Chase Bank            0   
23    JPMorgan Chase Bank            0   
24    JPMorgan Chase Bank            0   
25    JPMorgan Chase Bank            0   
26    JPMorgan Chase Bank            0   
27    JPMorgan Chase Bank            0   
28    JPMorgan Chase Bank            0   
29    JPMorgan Chase Bank            0   
...                   ...          ...   
5383  JPMorgan Chase Bank            0   
5384  JPMorgan Chase Bank            0   
5385  JPMorgan Chase Bank            0   
5386  JPMorgan Chase Bank            0   
5387  JPMorgan Chase Bank            0   
5388  JPMorgan Chase Bank            0   
5389  JPMorgan Chase Bank            0   
5390  JPMorgan Chase Bank            0   
5391  JPMorgan Chase Bank            0   
5392  JPMorgan Chase Bank            0   
5393  JPMorgan Chase Bank            0   
5394  JPMorgan Chase Bank            0   
5395  JPMorgan Chase Bank            0   
5396  JPMorgan Chase Bank            0   
5397  JPMorgan Chase Bank            0   
5398  JPMorgan Chase Bank            0   
5399  JPMorgan Chase Bank            0   
5400  JPMorgan Chase Bank            0   
5401  JPMorgan Chase Bank            0   
5402  JPMorgan Chase Bank            0   
5403  JPMorgan Chase Bank            0   
5404  JPMorgan Chase Bank            0   
5405  JPMorgan Chase Bank            0   
5406  JPMorgan Chase Bank            0   
5407  JPMorgan Chase Bank            0   
5408  JPMorgan Chase Bank            0   
5409  JPMorgan Chase Bank            0   
5410  JPMorgan Chase Bank            0   
5411  JPMorgan Chase Bank            0   
5412  JPMorgan Chase Bank            0   

                                      Branch Name  Branch Number  \
0                 JPMorgan Chase Bank Main Office              0   
1                   Vernon Hills Scarsdale Branch              2   
2            Great Neck Northern Boulevard Branch              3   
3                          North Hartsdale Branch              4   
4                        Lawrence Rockaway Branch              5   
5                             Mount Vernon Branch              9   
6                              Castle Hill Branch             12   
7                          Burnside Avenue Branch             15   
8                                 Gun Hill Branch             16   
9                         Kingsbridge Road Branch             18   
10                               Riverdale Branch             20   
11                      Washington Heights Branch             21   
12                                Broadway Branch             22   
13                      Westchester Square Branch             24   
14                 Bay Parkway and 85th St Branch             26   
15                        Grand Army Plaza Branch             33   
16                         Carroll Gardens Branch             34   
17              Third Avenue & 51st Street Branch             37   
18                                East End Branch             38   
19                                 Audubon B